color_refinement.py

```diff
@@ -183,6 +184,153 @@ def my_test(cg):
     fast_color_refine(cg)
 
 
+def choose_a_root(g: Graph) -> Vertex:
+    """
+    Determines the root for a tree that cuts the tree most in half, which is unique for a tree, so this should
+    result in the same root for both tree
+    :param g: Graph
+    :return: Root
+    """
+    # Choose a vertex to be the root (arbitrarily)
+    arb_root = g.vertices[0]
+    # Assign weights of the induced subgraphs
+    set_weight(arb_root)
+    # Retrieve the root by shifting
+    return shift(arb_root, g.order)
+
+
+def set_weight(root: Vertex, parent: Vertex = None):
+    """
+    Recursively assigns weights to the induced subgraphs starting with the root
+    :param root: Vertex
+    :param parent: The parent of the root, because those do not count in the weight of a subgraph
+    :return:
+    """
+    if root.degree == 1 and root.neighbours[0] == parent:
+        root.weight = 1
+        return 1
+    else:
+        for n in root.neighbours:
+            if n != parent:
+                root.weight += set_weight(n, root)
+        root.weight += 1
+        return root.weight
+
+
+def shift(vertex: Vertex, amount_verts: int) -> Vertex:
+    """
+    Returns the Vertex that is the root for this tree
+    :param vertex: the vertex we want to shift
+    :param amount_verts: the amount of vertices in the graph [invariant]
+    :return:
+    """
+    # If no neighbour of u has weight > n/2, return u
+    result = vertex
+    for n in vertex.neighbours:
+        if n.weight > amount_verts/2:
+            vertex.weight = vertex.weight - n.weight
+            n.weight = vertex.weight + n.weight
+            result = shift(n, amount_verts)
+    return result
+
+
+def assign_levels(root: Vertex, parent: Vertex = None, level: int = 0):
+    """
+    Assigns to a vertex the depth (in the algorithm called level, however our root = 0 instead of the max level)
+    :param root: The root of the tree
+    :param parent: The parent of the root (for the root of the tree this is None)
+    :param level: The level that needs to be assigned to the root
+    :return: Nothing because everything is assigned to the vertices
+    """
+    if root.level is None:
+        root.level = level
+    level += 1
+    for n in root.neighbours:
+        if n.level is None:
+            n.level = level
+        if n != parent:
+            root.children.append(n)
+            assign_levels(n, root, level)
+
+
+def initialize_tree(g: Graph):
+    """
+    The tree isomorphic algorithm and the root detection need certain attributes that are initialized here
+    :param g: Graph
+    :return: Nothing because everything is assigned to the vertices
+    """
+    for v in g.vertices:
+        v.weight = 0
+        v.level = None
+        v.children = []
+        v.value = None
+        v.tuples = []
+        # Assign all leaves integer 0
+        if v.degree == 1:
+            v.value = 0
+    return g
+
+
+def tree_isomorphism(g: Graph, h: Graph) -> bool:
```

Comment:
kunnen we ook checken hoeveel er zijn?

Reply:
Daar heb ik nog niets over kunnen vinden

Reply:
Mm moeten we daar dan de normale color refine voor gebruiken? Dat zou jammer zijn.

Reply:
Kun je niet iets doen met de tupels die je gegeven hebt?

Reply:
maar als je ziet dat ze isomorph zijn, dan kan je toch gewoon het aantal automorphisms tellen van 1 graph? Boom: isomorphism problem solved.